### ISSUE TITLE
Long time for prediction

### ISSUE THREAD

**Issue by skrishnanv** [MIC-DKFZ/nnUNet#417]
Hi,
After training a model, i am predicting for some datasets. Details:
image size: (512,512,608)
command used: nnUNet_predict --num_threads_preprocessing 1 --num_threads_nifti_save 1  -i ZA-16/nifti -o /tmp/ -t 402 -m 3d_lowres -chk model_best --disable_tta
time taken: 7 minutes, there are two volumes
CPU RAM: 24GB
GPU RAM: 16GB
What i have observed is that the following takes a lot of time:
force_separate_z: None interpolation order: 1
separate z: False lowres axis None
Is this expected behaviour?
Thank you

**Comment by harsharaman** [comment]
Same issue here

**Comment by FabianIsensee** [comment]
It is normal. Please make sure it is still running. If you are OOM your system will kill background workers and then it can get stuck. Also see documentation https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/common_problems_and_solutions.md

**Comment by Sungden** [comment]
hi, Fabian, I have the same question.   I am using this code on Task056 Verse segmentation. But on the inference stage ,it is very slow.(about 6 hours for one case(size 256x256x256)). I want to know it is normal? My cudnn version is 7.6.5. 
Thanks

**Comment by Sungden** [comment]
![time](https://user-images.githubusercontent.com/54644867/101320484-cbcb0400-389e-11eb-8066-fbc003e6c8c7.PNG)

**Comment by Sungden** [comment]
I used this command: CUDA_VISIBLE_DEVICES=2 nohup nnUNet_predict -i input_folder -o out_folder -m 3d_fullres -t Task056_VerSe -chk model_latest .

It has 26 labels.

**Comment by FabianIsensee** [comment]
I have worked with verse myself and it is not slow at all. Yes the segmentation export takes a while because of all the labels but it's much faster than 6 hours. I don't know why it is so slow for you

**Comment by Sungden** [comment]
When I used only a single model for inference ,the time is around 16 minutes, which I think is normal.
![t](https://user-images.githubusercontent.com/54644867/101346346-5de50380-38c3-11eb-9f80-7cbe5cf24433.PNG)

**Comment by FabianIsensee** [comment]
It is and it should not be this way. I don't know why this takes so long for you. Maybe your RAM is full and your PC is swapping?

**Comment by Sungden** [comment]
> It is and it should not be this way. I don't know why this takes so long for you. Maybe your RAM is full and your PC is swapping?

Yeah. I changed a new GPU machine and the time is normal now.
![time](https://user-images.githubusercontent.com/54644867/102352247-8915d400-3fe2-11eb-870a-6c53cd97cc5d.JPG)


Thanks!

**Comment by Sungden** [comment]
Hi, Fabian. I seem to know the real promblem why the inference time is so long for me because I  came across the following issue. 
![image](https://user-images.githubusercontent.com/54644867/102465629-dc426200-4088-11eb-9824-a5a3b924d0b6.png)

In the code below, you said that a npy file will be saved and then be read (and finally deleted) by the Process. In my case, the npy file was indeed saved and finally deleted, but there was no corresponding nii format file generated. 
![image](https://user-images.githubusercontent.com/54644867/102465879-2c212900-4089-11eb-8350-bdba85a3c9c7.png)

And some errors will appear below.
![image](https://user-images.githubusercontent.com/54644867/102466683-4d364980-408a-11eb-99eb-59cfd426e59f.png)

This is the usage situation of my PC.
![image](https://user-images.githubusercontent.com/54644867/102466843-87075000-408a-11eb-945d-042fa41e6ec2.png)

Thanks for your help.

**Comment by FabianIsensee** [comment]
Sounds like you are out of memory (RAM) to me. Or your disk is full
Best,
Fabian

**Comment by Sungden** [comment]
But I am using a total new machine. Could I send you several data for tesing?

**Comment by FabianIsensee** [comment]
How much RAM do you have?

**Comment by Sungden** [comment]
It is 32G. My data size is around 512x512x600.(not the VerSe challenge
dataset, it is the Task003_Liver dataset ).
[image: image.png]

So I plan to split a data into several small data or just use the 2D
network.

Thanks.

Fabian Isensee <notifications@github.com> 於 2020年12月18日週五 下午4:36寫道：

**Comment by FabianIsensee** [comment]
Yeah that's not enough. You can just create a swap partition on your ssd. This should be OK for inference

**Comment by Sungden** [comment]
Thanks!

**Comment by Sungden** [comment]
> Yeah that's not enough. You can just create a swap partition on your ssd. This should be OK for inference

Hi, Fabian. When I use a 128G RAM. The same problem still happened. I am sure it is the RAM out of memory. It seems nnUnet needs large memory (RAM) to process the segmentation of spine. Do you have any suggestion?
Thanks.

**Comment by FabianIsensee** [comment]
try `nnUNet_predict [...] --mode fast` or `nnUNet_predict [...] --mode fastest`
Segmentation quality is a little bit reduced but it requires less ram.
The problem is the interpolation of the segmentations to the original image spacing. This is unfortunately something I have limited control over.
Best,
Fabian

**Comment by Sungden** [comment]
thanks

**Comment by TunguskOdin** [comment]
Hello guys and Hello and THANK YOU to Dr.(already?) Isensee! I'm sure you already know your nnUNet is a BEAST and that the beautiful mountains near you can't wait until you go on vacation ;)

I know this is closed issue but I didn't want to open a new one about the same issue...

tl;dr- Is 12 minutes per CT scan for kidney segmentation (anywhere from 300 to 1300 slices) the best I can do with nnUNet_predict --disable_tta?

My inference/prediction times are also taking around 12.3 minutes for CT scans (usually of neck to thighs) for Kidney Tumor segmentation (2 labels). I downloaded the pretrained KiTS data and have run inference on the KiTS data as well as my own institute's set. The segmentations aren't perfect for some tumors and are pretty good for others. I come from the radiomics/manual segmentation side of things so your tool would be a great help!

I am already doing predictions with tta disabled and saw that Sungden got his inference times down to 1-2 and got jealous :)

I am worried I am under-utilizing my hardware because I have a setup with:
OS: Ubuntu 18.04
Conda env specifically for nnUNet
GPU: single card with 48 GBs VRAM (RTX 8000, Turing)
RAM: 192 GB
CPU: Intel 6226R (16 core, 64 'virtual' cores seen on htop)
Storage: 2x 1TB NVMe drives, RAID HDD and network storage (neither being used for inference currently)

I think I am under-utilizing my hardware because VRAM usage plateaus at around 2 GB (while GPU-Util is around 87-100%) and only 2-6 of the CPU virtual 'cores' are being used. RAM use ranges between 50-120 GB (running regular nnunet_predict with only --disable_tta flag)

I have also reached the OOM issue and end up with .npy files from 2-6GB in my output folder (I do not have a swap, so I will try adding swap for RAM overflow to fix this soon. this is happening for the images with the larger amount of slices, but those might not fit my use-case anyways)

In previous sessions, I have increased the num_threads_(preprocessing || nifti_save) to 12 and 4 but that did not increase speeds - only RAM usage)

I am also on CUDA 11.2, libcudnn 8.1.1, nvidia driver 460.32. pytorch was not compiled from source, but installed via conda.

I checked the benchmark page and for training am running only slightly better than the 2080Ti at ~90 seconds for Task 2 2D.

My question then: Is 12 minutes per CT scan the best I can do or have I provided enough room to improve?

Thank you, and if this should be its own issue/discussion post, please let me know!

**Comment by FabianIsensee** [comment]
Haha that must be the funniest post I have read in a long while :-D

Let me try and help you out. There are some things you can do. But first things first, let me answer and comment on your post:

> I think I am under-utilizing my hardware because VRAM usage plateaus at around 2 GB (while GPU-Util is around 87-100%) and only 2-6 of the CPU virtual 'cores' are being used. RAM use ranges between 50-120 GB (running regular nnunet_predict with only --disable_tta flag)
>In previous sessions, I have increased the num_threads_(preprocessing || nifti_save) to 12 and 4 but that did not increase speeds - only RAM usage)

Inference time is (most of the time) bottlenecked by the computing speed of your GPU. As long as there is sufficient CPU and RAM (and enough workers configured), these will not influence the inference speed substantially. Also don't get fooled by the GPU memory usage. It's just like with system RAM: as long as there is enough and the GPU is well utilized (high power consumption and GPU util) there is nothing to worry about.

> I am also on CUDA 11.2, libcudnn 8.1.1, nvidia driver 460.32. pytorch was not compiled from source, but installed via conda.
> I checked the benchmark page and for training am running only slightly better than the 2080Ti at ~90 seconds for Task 2 2D.

This is the first thing I would fix. Current pytorch build is broken. It does not accelerate 3D convs with tensorcores. 2D convs work though which is why your 2D benchmark indicates normal speed (you should have tested a 3D benchmark!). If you compile pytorch yourself (with the newest version of cuDNN) you will get ~2x speedup from that alone.

> I am already doing predictions with tta disabled and saw that Sungden got his inference times down to 1-2 and got jealous :)

Do not take inference time other people got on different tasks and data as reference. Each task has its own target spacing and image size which makes it difficult to compare. KiTS is one of the slowest datasets due to the high image resolution (spacing approx 0.8x0.8x0.8).

Here are some general remarks on inference that could help you:
- compile pytorch from source and verify that 3D benchmarks perform as expected
- consider using a single model for prediction rather than an ensemble (nnU-Net uses the 5 models from cross-validation as default)
- consider using the 3d_lowres config. It is much faster
- Recommendation: use nnUNet_predict with the following arguments: `--step_size 0.9 --all_in_gpu True --disable_tta True`. If you encounter Out of Memory Errors (GPU, not RAM!) then set `--all_in_gpu False`. I would also see if the speedup you get from other things you do will allow you to turn tta back on as this generally improves results.

Let me know if that helped.

Best,

Fabian

**Comment by FabianIsensee** [comment]
Just to give you a hint at the speed difference when compiling pytorch from source (these are epoch times during training):

pytorch 1.8.1 (and 1.8.1 nightly): **342s**
pytorch 1.7.1 compiled myself with cuDNN 8.1.0.77: **88.05s**

(gpu is RTX 3090)
Best,
Fabian

**Comment by hdnminh** [comment]
Hi @FabianIsensee 

How can I run pytorch 1.8 with nnUNet V2. Should I change here?
![image](https://github.com/MIC-DKFZ/nnUNet/assets/76483242/838bc983-74df-4150-ae43-72de22d8eba6)
